src/shipLogEntries.ts

```diff
@@ -180,6 +180,18 @@ export async function shipLogEntries(event: CloudWatchLogsEvent, context: Contex
         return log;
     });
     console.log(`Sending ${structuredLogs.length} events from ${logGroup} to ${kinesisStreamName} (with role: ${kinesisStreamRole})`);
-    
-    return await putKinesisRecords(kinesis, kinesisStreamName, structuredLogs);
+
+    const result = await putKinesisRecords(kinesis, kinesisStreamName, structuredLogs);
+
+    // If total failed record count across batches > 0,
+    // then raise an error so the invocation will be marked as failed
+    // and the client of lambda can respond as appropriate for them.
+    const failedRecordCount = result.reduce((count, output) => {
+        return count + (output.FailedRecordCount || 0)
+    }, 0);
+    if (failedRecordCount > 0) {
+        throw new Error(`failed to put ${failedRecordCount} record(s) on Kinesis stream`)
+    }
+
+    return result;
```

Comment:
Do we need a try...catch...finally or something to make the return happen? Or are we intentionally _not_ returning in the event of the `throw` happening?

Reply:
If you want to be purer you could return a `Promise.reject(...)` but otherwise JS will convert a thrown error into a rejected Promise due to the async function marker so it's kind of the same either way.